### PR TITLE
[Document] update assertion document

### PR DESCRIPTION
**Pull request by popcornylu**
Document the new metrics for assertions use.

1. Add new probability-based metric (`xxx_p`)
2. Change the text-length. (e.g. `min` -> `min_legnth` and `histogram` -> `histogram_length`)
3. Add assertion column to indicate if the metric is supported in assertion.
4. Update the diagram, the 100% is `sample count` instead of `row count`

Add new builtin and custom assertions document, which is originally documented in doc site
1. Rearrange the sections
2. add metric-based assertion
3. add new `assert_column_value` 

**PR checklist**

- [X] Ensure you have added or ran the appropriate tests for your PR.
- [X] DCO signed

**What type of PR is this?**
document

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
sc-28796

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
Yes

### CODE REVIEW

File: docs/metrics.md [hunk 1]
```diff
@@ -2,79 +2,93 @@
...
+| Column count              | Number of columns in this table                                                  | `col_count`    |           |         |
+| Sample count              | Number of rows been profiled. If no row limit is configured, the sample count is always the same as the row count. | `samples`      |           | 0.10.0  |
+| Sample percentage         | Percentage of rows been profiled                                                 | `samples_p`    |           | 0.11.0  |
+| Volume size *\**          | The volume size of this table in bytes                                           | `bytes`        | o         | 0.8.0   |
+| Created time *\**         | The time that this table created at in ISO 8601 format including time zone       | `created`      |           | 0.8.0   |
+| Last altered time *\**    | The last time that this table modified at in ISO 8601 format including time zone | `last_altered` |           | 0.8.0   |
+| Freshness *\**            | Time differentiation between the current time and table's last altered time      | `freshness`    | o         | 0.8.0   |
```
Comment (wcchang1115) [inline docs/metrics.md:14]: should we add that it's in second unit?

File: docs/assertions-builtin.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+</details>
+    
+
+## assert_column_value
+
+- Description: Assert the column value should be in the range.
+- Assert:
+    - `gte`:  the value should be greater than or equals to. (optional)
+    - `gt`:  the value should be greater than. (optional)
+    - `lte`:  the value should less than or equals to. (optional)
```
Comment (ctiml) [inline docs/assertions-builtin.md:45]:
```suggestion
    - `lte`:  the value should be less than or equals to. (optional)
```

File: docs/assertions-builtin.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+
+# Metric-based Assertion
+
+
+- Description: Metric-based assertions are assert the value of a metric.
+- Assert:
+    - `gte`:  the value should be greater than or equals to
+    - `gt`:  the value should be greater than
+    - `lte`:  the value should less than or equals to
```
Comment (ctiml) [inline docs/assertions-builtin.md:164]:
```suggestion
    - `lte`:  the value should be less than or equals to
```

File: docs/assertions-builtin.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    - `lt`:  the value should be less than
+    - `eq`:  the value should equals to
```
Comment (ctiml) [inline docs/assertions-builtin.md:166]:
```suggestion
    - `eq`:  the value should equal to
```

File: docs/assertions-custom.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
+# User defined test function
+
+In addition to builtins assertion function, PipeRider also supports adding customized assertion function. User defined
```
Comment (wcchang1115) [inline docs/assertions-custom.md:3]: should we change to assertion class or instance?

File: docs/assertions-custom.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    return result
+```
+
+If you have found something wrong, tried to add the reason to the result object:
+
+```python
+    def validate(self, context: AssertionContext) -> ValidationResult:
+        result = ValidationResult(context)
+        result.errors.append('explain to users why this broken')
+        return result
```
Comment (ctiml) [inline docs/assertions-custom.md:208]:
```suggestion
  def validate(self, context: AssertionContext) -> ValidationResult:
    result = ValidationResult(context)
    result.errors.append('explain to users why this broken')
    return result
```

File: docs/assertions-custom.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+# drwxr-xr-x  26 piperider  staff   832B  6  1 09:35 ..
+# -rw-r--r--   1 piperider  staff   6.2K  6  1 09:35 .profiler.json
+# -rw-r--r--   1 piperider  staff    15K  6  1 09:35 data.json
+```
+
+#### Validation
+
+Validation is used when `diagnose` or before `profiling` to verify the `assert` are well-defined in the assertion file.
+
+If everything was fine, the `validate` could return an ValidationResult without errors
```
Comment (ctiml) [inline docs/assertions-custom.md:194]:
```suggestion
If everything was fine, the `validate` could return an ValidationResult without errors:
```

File: docs/assertions-custom.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+return context.result.success() 
+```
+
+```python
+return context.result.fail()
+```
+
+## Profiling Data
+
+You might want to know what kinds of data you have in the `metric` dict? There is a `.profiler.json` in the output path
```
Comment (wcchang1115) [inline docs/assertions-custom.md:178]: need to change `.profiler.json` to `run.json`

File: docs/metrics.md [hunk 9]
```diff
@@ -2,79 +2,93 @@
...
-| Average | The average of a column. | integer, numeric | `avg` |  |
-| Sum | The sum of a column. | integer, numeric | `sum` |  |
-| Standard deviation | The standard deviation of a column. | integer, numeric, | `stddev` | 0.4.0 |
+| Metric             | Description                         | Column Type                | Field    | Assertion | Since |
+| ------------------ | ----------------------------------- | -------------------------- | -------- | :-------: | ----- |
+| Min                | The value of the minimum item       | integer, numeric, datetime | `min`    | o         |       |
+| Max                | The value of the maximum item       | integer, numeric, datetime | `max`    | o         |       |
+| Average            | The average of a column.            | integer, numeric           | `avg`    | o         |       |
+| Sum                | The sum of a column.                | integer, numeric           | `sum`    | o         |       |
+| Standard deviation | The standard deviation of a column. | integer, numeric,          | `stddev` | o         | 0.4.0 |
```
Comment (ctiml) [inline docs/metrics.md:80]:
```suggestion
| Standard deviation | The standard deviation of a column. | integer, numeric           | `stddev` | o         | 0.4.0 |
```

File: docs/assertions-custom.md (hunk 5 above)
Comment (popcornylu) [reply docs/assertions-custom.md:3]: won't fix

**Pull request merged**